File: 105. Construct Binary Tree from Preorder and Inorder Traversal/Solution.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder == null || inorder == null)
            return null;
        
        int preLen = preorder.length - 1;
        int inLen = inorder.length - 1;
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i <= inLen; i++)
            map.put(inorder[i], i);
    
        return dfs(map, preorder, 0, preLen, inorder, 0, inLen);
    }
    
    private TreeNode dfs(HashMap<Integer, Integer> map, int[] preorder, int preL, int preR, int[] inorder, int inL, int inR) {
        if (inL > inR || preL > preR)
            return null;
        
        TreeNode root = new TreeNode(preorder[preL]);
        int rootIndex = map.get(preorder[preL]);
        
        root.left = dfs(map, preorder, preL+1, rootIndex+preL-inL, inorder, inL, rootIndex-1);
        root.right = dfs(map, preorder, preL+rootIndex+1-inL, preR, inorder, rootIndex+1, inR);
        return root;
    }
}

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {integer[]} preorder
    # @param {integer[]} inorder
    # @return {TreeNode}
    def buildTree(self, preorder, inorder):
        if preorder == None or inorder == None:
            return None
            
        dict_tree = {}
        for i in xrange(len(inorder)):
            dict_tree[inorder[i]] = i
        return self.dfs(dict_tree, preorder, 0, len(preorder) - 1, inorder, 0, len(inorder) - 1)
        
    def dfs(self, dict_tree, preorder, preL, preR, inorder, inL, inR):
        if preL > preR or inL > inR:
            return None
        
        root = TreeNode(preorder[preL])
        rootIndex = dict_tree[preorder[preL]]
        
        root.left = self.dfs(dict_tree, preorder, preL+1, rootIndex+preL-inL, inorder, inL, rootIndex-1);
        root.right = self.dfs(dict_tree, preorder, preL+rootIndex+1-inL, preR, inorder, rootIndex+1, inR);
        return root